### PR TITLE
[Llama FA2] Re-add _expand_attention_mask and clean a couple things

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR cleans the attention mask converter a bit more, corrects some docstrings and removes outdated comments and deprecates `_expand_attention_mask` to fix optimum.

### CODE REVIEW

File: src/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -827,10 +837,9 @@ def _init_weights(self, module):
             if module.padding_idx is not None:
                 module.weight.data[module.padding_idx].zero_()
 
-    def _set_gradient_checkpointing(self, module, gradient_checkpointing_func=None):
+    def _set_gradient_checkpointing(self, module, value=False):
         if isinstance(module, LlamaModel):
```
Comment (patrickvonplaten) [inline src/transformers/models/llama/modeling_llama.py:841]:
```suggestion
        if isinstance(module, LlamaModel):
            module.gradient_checkpointing_func = gradient_checkpointing_func
            module.gradient_checkpointing = gradient_checkpointing_func is not None
```

File: src/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -827,10 +837,9 @@ def _init_weights(self, module):
             if module.padding_idx is not None:
                 module.weight.data[module.padding_idx].zero_()
 
-    def _set_gradient_checkpointing(self, module, gradient_checkpointing_func=None):
+    def _set_gradient_checkpointing(self, module, value=False):
```
Comment (patrickvonplaten) [inline src/transformers/models/llama/modeling_llama.py:840]:
```suggestion
    def _set_gradient_checkpointing(self, module, gradient_checkpointing_func=None):
```

File: src/transformers/models/llama/modeling_llama.py [hunk 3]
```diff
@@ -1014,14 +1023,16 @@ def forward(
...
+
+                def create_custom_forward(module):
+                    def custom_forward(*inputs):
+                        # None for past_key_value
+                        return module(*inputs, past_key_value, output_attentions)
+
+                    return custom_forward
+
+                layer_outputs = torch.utils.checkpoint.checkpoint(
+                    create_custom_forward(decoder_layer), hidden_states, attention_mask, position_ids
```
Comment (patrickvonplaten) [inline src/transformers/models/llama/modeling_llama.py:1035]:
```suggestion
                layer_outputs = self.gradient_checkpointing_func(
                    decoder_layer.__call__,
                    hidden_states,
                    attention_mask,
                    position_ids,
                    past_key_value,
                    output_attentions,
                    use_cache,
```

File: src/transformers/models/llama/modeling_llama.py [hunk 4]
```diff
@@ -871,7 +882,7 @@ def _set_gradient_checkpointing(self, module, gradient_checkpointing_func=None):
         past_key_values (`tuple(tuple(torch.FloatTensor))`, *optional*, returned when `use_cache=True` is passed or when `config.use_cache=True`):
             Tuple of `tuple(torch.FloatTensor)` of length `config.n_layers`, with each tuple having 2 tensors of shape
             `(batch_size, num_heads, sequence_length, embed_size_per_head)`) and 2 additional tensors of shape
-            `(batch_size, num_heads, decoder_sequence_length, embed_size_per_head)`.
+            `(batch_size, num_heads, encoder_sequence_length, embed_size_per_head)`.
```
Comment (patrickvonplaten) [inline src/transformers/models/llama/modeling_llama.py:885]:
```suggestion
            `(batch_size, num_heads, decoder_sequence_length, embed_size_per_head)`.
```

File: src/transformers/models/llama/modeling_llama.py [hunk 5]
```diff
@@ -831,6 +841,7 @@ def _set_gradient_checkpointing(self, module, gradient_checkpointing_func=None):
         if isinstance(module, LlamaModel):
             module.gradient_checkpointing_func = gradient_checkpointing_func
             module.gradient_checkpointing = gradient_checkpointing_func is not None
+            module.gradient_checkpointing = value
```
Comment (patrickvonplaten) [inline src/transformers/models/llama/modeling_llama.py:844]:
```suggestion
```

**Comment by patrickvonplaten** [comment]
@ArthurZucker could you give this a quick review? It'd make my Bart FA PR much nicer and should also fix the better transformers problem with optimum

**Comment by ArthurZucker** [comment]
Of course!

**Review by ArthurZucker — APPROVED** [review]

File: src/transformers/models/llama/modeling_llama.py [hunk 6]
```diff
@@ -740,7 +752,8 @@ def forward(
         Args:
             hidden_states (`torch.FloatTensor`): input to the layer of shape `(batch, seq_len, embed_dim)`
             attention_mask (`torch.FloatTensor`, *optional*): attention mask of size
```
Comment (patrickvonplaten) [inline src/transformers/models/llama/modeling_llama.py:754]:
```suggestion
            attention_mask (`torch.FloatTensor`, *optional*):
```

File: src/transformers/models/llama/modeling_llama.py [hunk 7]
```diff
@@ -740,7 +752,8 @@ def forward(
...
-                `(batch, sequence_length)` where padding elements are indicated by 0.
+                `(batch_size, sequence_length)` if flash attention is used or `(batch_size, 1, query_sequence_length,
```
Comment (patrickvonplaten) [inline src/transformers/models/llama/modeling_llama.py:755]:
```suggestion
                attention mask of size `(batch_size, sequence_length)` if flash attention is used or `(batch_size, 1, query_sequence_length,
```

File: src/transformers/models/llama/modeling_llama.py [hunk 8]
```diff
@@ -64,6 +64,13 @@ def _get_unpad_data(attention_mask):
     )
 
 
+def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int] = None):
+    warnings.warn(
+        "Calling `transformers.models.llama.modeling_llama._expand_mask` is deprecated and will be removed in v4.37. Use `transformers.models.llama.modeling_llama.AttnMaskConverter._expand_mask"
+    )
+    return AttnMaskConverter._expand_mask(mask=mask, dtype=dtype, tgt_len=tgt_len)
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/transformers/models/llama/modeling_llama.py:71]: Think in optimum only the llama mask utils are imported: https://github.com/huggingface/optimum/blob/313e1bd0de2b44aaa71797464f1e8b6a041a6f18/optimum/bettertransformer/models/attention.py#L25

[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/llama/modeling_llama.py:71]: ok 👍🏻

File: src/transformers/models/llama/modeling_llama.py [hunk 9]
```diff
@@ -740,7 +752,8 @@ def forward(
         Args:
             hidden_states (`torch.FloatTensor`): input to the layer of shape `(batch, seq_len, embed_dim)`
             attention_mask (`torch.FloatTensor`, *optional*): attention mask of size
-                `(batch, sequence_length)` where padding elements are indicated by 0.
+                `(batch_size, sequence_length)` if flash attention is used or `(batch_size, 1, query_sequence_length,
+                key_sequence_length)` if default attention is used.
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/transformers/models/llama/modeling_llama.py]: Thanks!
